docs/plutus-report/scripting.tex

```diff
@@ -111,7 +115,127 @@ \subsection{Recursive values}
+The key issue is that we may want to have \emph{different} sets of builtins in different situations.
+For example:
+\begin{enumerate}
+\item
+  During testing it can be convenient to have a builtin which implements \emph{tracing}: emitting log lines to a global trace output.
+  But we don't want this functionality on-chain, so we want to remove the builtin (or replace it with a no-op).
+\item
+  We may need different builtins for different ledgers (see \cref{req:script-lang-extensibility}).
+\end{enumerate}
+
+So we need to make builtin types and functions extensible.
+
+\subsubsection{Builtin types}
+
+Extensible builtin types are relatively straightforward on the type level, but become trickier on the value level.
+Values of builtin types are \emph{constants}, but what is a constant?
+It's easy to say what a \code{Integer} constant should be: an integer!
+However, what about constants of extensible builtin types?
+
+We adopt an approach familiar from generic programming in dependently-typed languages, where we parameterize our language over a \emph{universe} of types.\footnote{
+  Where do these types themselves come from?
+  We assume that we are working in some ``metalangauge'' which has types such as \code{Integer} that we can reuse in Plutus Core.
+}
+A constant then is then \emph{tagged} with a value corresponding to a type in that universe, and a value of that type.
+
+This does complicate interpretation of \gls{plutus-core} programs, since in order to interpret a program, you also need to know what universe of types it is expecting.
+In practice this is not so bad: for most users (e.g. the \gls{cardano} ledger) there will be a single universe that they use, and from there on they can simply act as though we did not have extensible builtins.
+
+\paragraph{Builtin datatypes}
+
+Adding new types to \gls{plutus-core} is not very useful unless they have functions that can operate on them!
+A particularly common case is where the new type is (morally) a datatype, and should support pattern matching.
+We can support this by also providing a ``matching'' function as a builtin function.\footnote{
+  These functions look just like the ``matchers'' for Scott-encoded datatypes, which is no coincidence.
+}
+
+For example, the matching function for \code{bool} is \code{ifThenElse :: forall a . bool -> a -> a -> a}.
+Matching functions are always polymorphic, which is a key reason why we need to support polymorphic builtin functions (see \cref{sec:polymorphic-builtins}).
+
+\paragraph{Polymorphic builtin types}
+
+It would be nice to support polymorphic builtin types.
+For example, that would allow us to provide optimized implementations of container types such as maps.
+
+This seems feasible in principle, but we have not worked the details out yet.
+
+\subsubsection{Builtin functions}
+
+We could implement extensible builtin functions in a similar way to extensible builtin types: use some universe that catalogues the available builtin functions.
```

Comment:
I'd avoid applying the "universe" term to functions (or anything else living at the term level). "Universe" commonly denotes a set of types (regardless of whether it's a close universe of built-in types like we have or an open one). I don't know if it's incorrect or not, but it reads a little weirdly to me.

Reply:
Okay, I'll use some other word.